### PR TITLE
[WIP] Superstatic 4.0

### PR DESCRIPTION
**Pull request by mbleigh**
Version 3.0 of Superstatic was written (a bit hastily) to make Divshot and Firebase configs compatible with each other as much as possible. There was still significant work to be done in order to complete the transition from Divshot to Firebase and make it possible for Superstatic to fully power Firebase Hosting's back-end.

This PR is a substantial refactor of pretty much everything in Superstatic and how it all works. The following are an incomplete set of breaking changes for 4.0:

* All config will be camel-cased moving forward and will generally match Firebase Hosting's [config](https://www.firebase.com/docs/hosting/guide/full-config.html).
* The package is being slimmed down as much as possible -- LiveReload is being removed, small packages that can be inlined are, etc.
* The responder has been completely rewritten. It now lives at `res._responder` and has promise-based clever logic to do things like try multiple files and fall back to `next()` if none are found.
* Superstatic services are removed. The plan is to make it possible to inject standard middleware before or after any of the Superstatic processing steps. These middleware will not be baked into the Superstatic command-line runner.
* The provider system has been completely rewritten. A provider is now just a simple Promise-returning function that takes a path and returns a payload including a stream, modified time, size, and etag.

These changes are sweeping but necessary for the library to stay healthy and actively developed moving forward. Firebase Hosting will be running version 4.0 (once it's ready).

### Todo

* [x] Hunt down and replace all instances of underscored configuration
* [x] Rip out the logic that supports both object and array styles in rewrites and redirects
* [x] Update README to reflect changes to config styles
* [x] Update README with how to create a simple provider
* [x] Create a `memory` provider for easy in-memory testing etc
* [x] Look for opportunities to remove dependencies and do so
* [x] Add a "strict" mode that will only support things that Firebase Hosting explicitly supports
* [x] Add istanbul test coverage, switch from JSHint to ESLint

### CODE REVIEW

**Comment by mbleigh** [comment]
@drtriumph FYI, will probably have you review when it's totally ready.

File: README.md [hunk 1]
```diff
@@ -96,28 +96,28 @@ apps. An example:
 }
 ```
 
-**redirects:** you can specify certain url paths to be redirected to another url by supplying configuration to the `redirects` key. Path matching is similar to using custom routes. Redirects use the `301` HTTP status code by default, but this can be overridden by configuration.
+**function():** you can specify certain url paths to be redirected to another url by supplying configuration to the `function()` key. Path matching is similar to using custom routes. function() use the `301` HTTP status code by default, but this can be overridden by configuration.
```
Comment (drtriumph) [inline README.md]: Is the change from `"redirects"` to `"function()"` intentional or just a find/replace error?

File: lib/activator.js [hunk 2]
```diff
@@ -41,7 +41,7 @@ module.exports = function(spec, provider, router) {
         stack.push(middleware.services(spec));
       }
 
-      // Redirects
+      // function()
```
Comment (drtriumph) [inline lib/activator.js]: And again here

File: lib/assets/env.js.template [hunk 3]
```diff
@@ -13,6 +13,6 @@
         // Browser globals
         root.__env = factory();
     }
-}(this, function () {
+}(this, function() {
     return {{ENV}};
 }));
```
Comment (drtriumph) [inline lib/assets/env.js.template]: need a newline here

File: lib/cli/index.js [hunk 4]
```diff
@@ -37,7 +36,7 @@ module.exports = function () {
   //       server: require('../server')
   //     }
   //   }
-  // ], function (err) {});
+  // ], function(err) {});
```
Comment (drtriumph) [inline lib/cli/index.js]: nit: commented section can be removed

File: lib/middleware/static-router.js [hunk 5]
```diff
@@ -4,18 +4,17 @@
...
-var url = require('fast-url-parser');
-var flattenToObject = require('flatten-to-object');
+var urlParser = require('fast-url-parser');
 var path = require('path');
 var _ = require('lodash');
 
 var normalizeConfig = function(config) {
-  if (!config) return [];
-  if (!_.isArray(config)){ config = [config]; }
+  if (!config) { return []; }
```
Comment (drtriumph) [inline lib/middleware/static-router.js]: nit: these could be on their own lines

File: lib/utils/pathutils.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+var join = require('join-path');
+var _ = require('lodash');
+
+var INDEX_FILE = 'index.html';
+
+var pathutils = {
+  asDirectoryIndex: function(p) {
```
Comment (drtriumph) [inline lib/utils/pathutils.js]: nit: not generally a fan of single-char variables. `path` isn't too overloaded here

**Comment by drtriumph** [comment]
Just a few nits and one comment about the README / comments, but otherwise this is looking great. I particularly love the negative lines added

**Comment by mbleigh** [comment]
This has been reviewed in incremental PRs by @drtriumph along the way, forgot to ask him to merge and he's on UK time so I'm self-merging.

**Pull request merged**